improver/utilities/extract_from_table.py

```diff
@@ -0,0 +1,175 @@
+
+        result = table.to_numpy()[row_index, column_index]
+        result = result.reshape(shape)
+        return result
+
+    def convert_dict_to_dataframe(self, table: dict) -> DataFrame:
+        """Converts a dictionary to a pandas DataFrame"""
+
+        table = DataFrame.from_dict(table)
+        table.columns = table.columns.astype(float)
+        table.index = table.index.astype(float)
+
+        table = table.reindex(sorted(table.columns), axis=1)
+        table = table.reindex(sorted(table.index), axis=0)
+
+        return table
+
+    def process(self, *cubes: Union[iris.cube.CubeList, iris.cube.Cube], table: dict):
+        """
+        Process the input cubes and extract values from a table based on the provided row and
+        column indices. The row name is used to identify the cube used for indexing the rows
+        of the table. The new name is an optional argument that can be used to rename the
+        resulting cube.
+
+        Args:
+            cubes:
+                Input cubes for indexing columns and rows of the table. Exactly 2 cubes should
+                be provided one which contains the values to extract from the rows and one for
+                the columns.
+            table:
+                A dictionary representing the table from which values are extracted. Dictionary
+                should be in the form: {column_name_1:{row_name_1:value, row_name_2:value},...}
+
+        Returns:
+            Cube of the same shape and metadata as the row input cubes with values extracted
+            from the table based on the row and column input cubes. The cube will be re-named
+            if new_name is provided.
+
+        Raises:
+            ValueError:
+                - If exactly 2 cubes are not provided.
+                - If the shapes of the column and row cubes do not match.
+        """
+        cubes = as_cubelist(*cubes)
+
+        if len(cubes) != 2:
+            raise ValueError(
+                f"""Exactly 2 cubes should be provided, one for indexing columns and one for
+                indexing rows. Provided cubes are {[cube.name() for cube in cubes]}"""
+            )
+
+        row_cube = cubes.extract_cube(self.row_name)
+        cubes.remove(row_cube)
+        column_cube = cubes[0]
+        coord_order = [coord.name() for coord in column_cube.coords()]
+        enforce_coordinate_ordering(row_cube, coord_order)
+
+        if column_cube.shape != row_cube.shape:
+            raise ValueError(
+                f"""Shapes of cubes do not match. Column cube shape:
+                {column_cube.shape}, row cube shape: {row_cube.shape}"""
+            )
+
+        table = self.convert_dict_to_dataframe(table)
+        result = self.extract_table_values(table, column_cube, row_cube)
+        if result.dtype == np.float64:
+            result = result.astype(np.float32)
+        result_cube = row_cube.copy(data=result)
+        if self.new_name:
+            result_cube.rename(self.new_name)
```

Comment:
I think we need the ability to set the units too. Ideally, these would come from the table so that they are defined with the data they represent.

Reply:
I've added a metadata section to the dictionary which adds an option to define units and any other metadata we might want to include with a table.

Reply:
You can define other metadata, but it appears that you only make it possible to use the units. Is that enough for now?

Reply:
It's sufficient from an EPP perspective. The rest of the metadata such as the name was just so you can identify what the data is.